### PR TITLE
Feat: Added Pug parser with inital queries

### PR DESCRIPTION
**Pull request by zealot128**
I discussed my progress in #1520 

This is a WIP with the Pug tree-parser. If worked OK for my type of files. I am sure it does not support all of the syntax, yet, but it would be much better than a white screen.

Bildschirmfoto 2021-07-10 um 17 18 19

As you can see, the highlighting is quite ok for raw "example.pug" files.

Unfortunately, if does not work with Vue template files, yet. I've added the ``ftdetect`` rule, but it is unable to map the ``@pug`` with the pug file

```
<template lang="pug">
body: div 
  | this is unhighlighted
</template>
```

Is unhighlighted.... For test, I've changed the ``vue/injections.scm`` template rule from ``@pug`` to @typescript or @scss, and the highlighting works.... What am I missing? Where can I tell nvim how to handle the ``@pug`` marked parts as a pug file.

I've also tried using the @content + @language way described in the README without any success:

```
(template_element
 (start_tag
  (attribute
   ((attribute_name) @lang_key (#match? @lang_key "lang"))
   ((quoted_attribute_value) @language)
  )
 )
 ((text) @content)
 )
```

But it seems to be a problem with the vue-parer, which already processes the content of the script tag as html... If anybody has any ideas, I'll be glad!

### CODE REVIEW

**Comment by zealot128** [comment]
Finally, I've made it work with embedded pug inside vue.


I have still a PR on the tree-sitter-vue project. When that is merged, the pug highlighting within vue will also work!
https://github.com/ikatyang/tree-sitter-vue/pull/13

Bildschirmfoto 2021-07-12 um 00 06 44

File: lua/nvim-treesitter/parsers.lua
Comment (theHamsta) [inline lua/nvim-treesitter/parsers.lua:401]: Why `used_by` vue? This isn't necessary for language injection and would only have an effect if a vue parser wouldn't exist and the filetype should use the pug parser.

**Review by theHamsta — CHANGES_REQUESTED** [review]
Appart from the `used_by` this looks good

File: ftdetect/pug.vim
Comment (vigoux) [inline ftdetect/pug.vim:1]: I'd rather do
```suggestion
au BufRead,BufNewFile *.pug		setlocal filetype=pug
```

File: queries/pug/highlights.scm
Comment (vigoux) [inline queries/pug/highlights.scm:17]: You can directly do `constant.builtin` here rather than the `tag_name` proxy.

File: queries/pug/highlights.scm
Comment (vigoux) [inline queries/pug/highlights.scm:23]: Duplicate?

File: queries/pug/highlights.scm
Comment (vigoux) [inline queries/pug/highlights.scm:28]: Same here than above.

File: ftdetect/pug.vim
Comment (zealot128) [reply ftdetect/pug.vim:1]: Fine by me, I've just peeked into the existing ftdetect rules [ftdetect/gomod.vim](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/ftdetect/gomod.vim) and [ftdetect/gmd.vim](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/ftdetect/gdscript.vim) are doing the same with ``set`` instead of ``setlocal``. Maybe they should be changed too?

File: queries/pug/highlights.scm
Comment (zealot128) [reply queries/pug/highlights.scm:23]: My intention was, to make highlighting for "static" attributes different than those with Javascript interpolation.

HTML Attributes such as "v-if=" , "@click=", ":src=" are interpreted as Javascript within Vue template. That's why I used the second, more specific rule, to override the highlighting for those attribute. 
See here the "target" attribute is plain text as v-if & src are bound to Javascript values.

Bildschirmfoto 2021-07-13 um 18 56 01

File: queries/pug/highlights.scm
Comment (zealot128) [reply queries/pug/highlights.scm]: Great! committed and tested.

File: ftdetect/pug.vim
Comment (zealot128) [reply ftdetect/pug.vim]: changed it to setlocal, which should make much more sense for most of these autocomands. Like i said, maybe worth changing the other 2 ftdetect scripts the same way.

File: lua/nvim-treesitter/parsers.lua
Comment (zealot128) [reply lua/nvim-treesitter/parsers.lua]: removed it! as my prime goal was to get the pug highlighting within vue working I tried all avenues I could find :) In the end it is necessary to change the parser grammar within tree-sitter-vue to make it work, which I also submitted a [PR](https://github.com/ikatyang/tree-sitter-vue/pull/13). 
I removed the used_by line.

File: queries/pug/highlights.scm
Comment (zealot128) [reply queries/pug/highlights.scm]: You mean the proxy indirection? I've moved the @keyword into the parens.

File: ftdetect/pug.vim
Comment (vigoux) [reply ftdetect/pug.vim]: I agree on changing those we'll do it in another pr!

**Review by theHamsta — APPROVED** [review]

**Comment by zealot128** [comment]
About the Action failures...

I also got the same error ``./scripts/check-queries.lua:64: .../usr/share/nvim/runtime/lua/vim/treesitter/query.lua:161: query: error at position 737`` when opening any *.scm files on my local machine **after** I rebased onto origin/master.

The error is still present on my local machine, even if I check out origin/master without my changes, so I guess the error is something else. 

Only when I go BEFORE 2a0097828d35a170e43616777c06611d881324c9 then the error disappears, so I guess that commit seemed to have introduced the query problem which seems logical, as the query highlighting was touched.

**Comment by theHamsta** [comment]
The error is in the vue injections. Maybe the vue parser was updated?

How can the pug parser be installed without `requires_generate_from_source` when the repo does not provide `parser.c`?

**Comment by zealot128** [comment]
> The error is in the vue injections. Maybe the vue parser was updated?

The last update from tree-sitter-vue on master branch was from April, so no recent changes there. 
If I check out nvim-tressitter@master and revert the ``queries/query/highlights.scm`` back to c37e79803e21abfae960174a6c661da166c87e8b, than there is no error. So 2a0097828d35a170e43616777c06611d881324c9 introduced the error, which I receive the same when viewing any query file at all:

Bildschirmfoto 2021-07-14 um 08 55 10


> 
> How can the pug parser be installed without `requires_generate_from_source` when the repo does not provide `parser.c`?

My fork of the pug parser now includes all the bindings/*, parser.c, ... files to be able to install directly without having to run 'tree-sitter generate' before.
TSInstall pug as well as TSInstallFromGrammar pug worked on my machine.

**Comment by theHamsta** [comment]
@zealot128  It seems that it has to be `text` instead of `rawtext`
```
(
  (template_element
    (start_tag
      (attribute
        (quoted_attribute_value (attribute_value) @_lang)))
    (text) @pug)
  (#match? @_lang "pug")
)
```

**Comment by zealot128** [comment]
> @zealot128 It seems that it has to be `text` instead of `rawtext`
> 
> ```
> (
>   (template_element
>     (start_tag
>       (attribute
>         (quoted_attribute_value (attribute_value) @_lang)))
>     (text) @pug)
>   (#match? @_lang "pug")
> )
> ```

@theHamsta i thought so the same, but it didn't work. TreeSitterVue always parses the content of "<template>" as html so Pug would never have a Chance to handle the unmodified output via injection (VueTreeSitter does the same already for the <script> and <style> tags and provides the unmodified raw_text as target for injection)..

I opened a PR on the Vue repo which adds the "raw_text" node for the "template" tag when used in conjunction with an attribute.

https://github.com/ikatyang/tree-sitter-vue/pull/13

Is this the the reason the checks are failing here?

**Comment by theHamsta** [comment]
yes, tree-sitter can check whether `raw_text` is actually allowed at that position.

**Comment by zealot128** [comment]
Oh great, that means either wait until the other PR is merged, or I modify this PR first, that it does not touch Vue integration for now.

**Review by vigoux — APPROVED** [review]

**Comment by theHamsta** [comment]
@zealot128 your injection query should now be fixed

**Pull request merged**
